pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go

```diff
@@ -0,0 +1,289 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package assemble
+
+import (
+	runtimev1alpha1 "github.com/crossplane/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha2"
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	ctrlutil "github.com/oam-dev/kubevela/pkg/controller/utils"
+	"github.com/oam-dev/kubevela/pkg/oam"
+	"github.com/oam-dev/kubevela/pkg/oam/util"
+)
+
+// NewAssembleOptions create a Options
+func NewAssembleOptions(appRevision *v1beta1.ApplicationRevision) *Options {
+	return &Options{AppRevision: appRevision}
+}
+
+// Options contains configuration to assemble resources recorded in the ApplicationRevision.
+// 'Assemble' means completing workload/trait resources and get them ready to emit into K8s
+// cluster.
+type Options struct {
+	AppRevision     *v1beta1.ApplicationRevision
+	WorkloadOptions []WorkloadOption
+
+	AppConfig      *v1alpha2.ApplicationConfiguration
+	Comps          []*v1alpha2.Component
+	AppName        string
+	AppNamespace   string
+	AppLables      map[string]string
+	AppAnnotations map[string]string
+	AppOwnerRef    *metav1.OwnerReference
+
+	// map key is component name
+	AssembledWorkloads map[string]*unstructured.Unstructured
+	AssembledTraits    map[string][]*unstructured.Unstructured
+	ReferencedScopes   map[string][]runtimev1alpha1.TypedReference
+}
+
+// WorkloadOption will be applied to each workloads AFTER it has been assembled by generic rules shown below:
+// 1) use component name as workload name
+// 2) use application namespace as workload namespace if unspecified
+// 3) set application as workload's owner
+// 4) pass all application's lables and annotations to workload's
+// Component and ComponentDefinition are enough for caller to manipulate workloads.
+// Caller can use below lables of workload to get more information:
+// - oam.LabelAppName
+// - oam.LabelAppRevision
+// - oam.LabelAppRevisionHash
+// - oam.LabelAppComponent
+// - oam.LabelAppComponentRevision
+type WorkloadOption interface {
+	ApplyToWorkload(workload *unstructured.Unstructured, comp *v1alpha2.Component, compDefinition *v1beta1.ComponentDefinition) error
+}
+
+// WithWorkloadOption add a WorkloadOption
+func (o *Options) WithWorkloadOption(wo WorkloadOption) *Options {
+	if o.WorkloadOptions == nil {
+		o.WorkloadOptions = make([]WorkloadOption, 0)
+	}
+	o.WorkloadOptions = append(o.WorkloadOptions, wo)
+	return o
+}
+
+// Assemble an application's manifests including workloads and traits according to a specific application revision
+func (o *Options) Assemble() (map[string]*unstructured.Unstructured, map[string][]*unstructured.Unstructured, map[string][]runtimev1alpha1.TypedReference, error) {
```

Comment:
This is too tightly coupled to components and traits.
It should just take in appRev and return an array of Unstructureds.

Reply:
Yes, that's truly a problem... My concern is, is it possible to have a scenario where consumer of assemble want to distinguish these resource and do something based on them? If no, I think it's ok to merge them into an array of Unstructureds.

Reply:
Or provide two interfaces, one is to return all in one array, the other is to return categorized results.

Reply:
Let's return an array for now. If there are requests from the community for other results, we can add them at that point.

Reply:
Got it! Will do that.